### PR TITLE
[BE] add warning message to cmake against env var "-std=c++xx"

### PR DESCRIPTION
**Pull request by walterddr**
this was discovered when working on #50230.

environment variables such as CXXFLAGS="-std=c++17" will not work because we use CMAKE_CXX_STANDARD 14. 
Adding this warning to alert users when environment variable was set.

See: [CMake env var usage](https://cmake.org/cmake/help/latest/manual/cmake-env-variables.7.html#id4) and [CXXFLAGS usage](https://cmake.org/cmake/help/latest/envvar/CXXFLAGS.html) for more details.

### CODE REVIEW

**Review by samestep — APPROVED** [review]

**Comment by samestep** [comment]
test plan?

**Comment by walterddr** [comment]
it should be pretty obvious from the description? (setting CXXFLAGS with `-std=c++17` and it should produce a warning)  any suggestion to improve the description?

**Comment by samestep** [comment]
> it should be pretty obvious from the description? (setting CXXFLAGS with `-std=c++17` and it should produce a warning) any suggestion to improve the description?

right so could you add to the description an example command you could run that would give different output before vs after this PR, and show the before/after output?

**Comment by walterddr** [comment]
that would be pretty hard since it depends on what current user environment variable is. to fully explain with context would require the commit message to be quite large. 

However I will create an issue to explain the full context and link to this PR.

**Comment by samestep** [comment]
> that would be pretty hard since it depends on what current user environment variable is. to fully explain with context would require the commit message to be quite large.

oh, so it wouldn't be as simple as setting `CXXFLAGS="-std=c++17"` on the command line and then calling CMake? I guess I'm a bit how it can simultaneously "pretty obvious from the description" and also require a very long commit message to explain

**Comment by walterddr** [comment]
> > that would be pretty hard since it depends on what current user environment variable is. to fully explain with context would require the commit message to be quite large.
> 
> oh, so it wouldn't be as simple as setting `CXXFLAGS="-std=c++17"` on the command line and then calling CMake? I guess I'm a bit how it can simultaneously "pretty obvious from the description" and also require a very long commit message to explain

this won't work if you already have customized CXXFLAGS set for example CXXFLAGS="-Werror -fPIC -O2 ..." -... (essentially it removed all those CXXFLAGS in environment and set it only to "-std=c++17"). this wont work if we do CXXFLAGS="${CXXFLAGS} -std=c++17" either because originally CXXFLAGS might already contain -std=c++17. 

long story short the description can be as complicated as checking all environment varaibles and see if it has set cxx standard and if not put it in to produce the warning, and remove it to produce no warning..

hence in my description instead of putting a copy-paste command I put in "add -std=c++17 properly to your CXXFLAGS". to avoid users look back at this commit message in the future to blindly copy-paste the test plan

**Comment by samestep** [comment]
hmm I think we might have different definitions of "test plan"; I just mean some example command that you can run that demonstrates the new warning message you added, so for instance it sounds like you can just set `CXXFLAGS` to `-std=c++17` and then run CMake

if the test plan propagated whatever `CXXFLAGS` were already set, then it would actually be a worse test plan, because it would be less deterministic

**Comment by walterddr** [comment]
good point. adding links to cmake environment variable usage link to description

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,8 +30,16 @@ endif()
 
 set(CMAKE_INSTALL_MESSAGE NEVER)
 
+# check and set CMAKE_CXX_STANDARD
+string(FIND "${CMAKE_CXX_FLAGS}" "-std=c++" contains_manual_std_cpp)
+if(contains_manual_std_cpp GREATER -1)
```
Comment (walterddr) [inline CMakeLists.txt:35]:
```suggestion
string(FIND "${CMAKE_CXX_FLAGS}" "-std=c++" env_cxx_standard)
if(env_cxx_standard GREATER -1)
```

**Pull request closed without merging**